python/iceberg/parquet/parquet_rowgroup_evaluator.py

```diff
@@ -0,0 +1,236 @@
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+from iceberg.api.expressions import Binder, Expressions, ExpressionVisitors
+from iceberg.api.types import Conversions
+
+
+class ParquetRowgroupEvaluator():
+    """
+    Handles Reading a parquet file for a given start to end range.  Also, enforces
+    the projected iceberg schema on the read data.  This may involve selectively reading
+    columns, remapping column names, re-arranging the projection ordering, and possibly
+    adding null columns
+
+    Parameters
+    ----------
+    schema : iceberg.api.Schema
+        An iceberg schema to use for binding the predicate
+    unbound : iceberg.api.expressions.UnboundPredicate
+        The unbound predicate to evaluate
+    start : int
+        The start index of the assigned reader
+    end : int
+        The end-index of the assigned reader
+    """
+    def __init__(self, schema, unbound, start, end):
+        self.schema = schema
+        self.struct = schema.as_struct()
+
+        self.expr = None if unbound is None else Binder.bind(self.struct, Expressions.rewrite_not(unbound))
+        self.start = start
+        self.end = end
+        self._visitors = None
+
+    def _visitor(self):
+        if self._visitors is None:
+            self._visitors = ParquetRowgroupEvalVisitor(self.expr, self.schema, self.struct,
+                                                        self.start, self.end)
+
+        return self._visitors
+
+    def eval(self, row_group):
+        return self._visitor().eval(row_group)
+
+
+class ParquetRowgroupEvalVisitor(ExpressionVisitors.BoundExpressionVisitor):
+    ROWS_MIGHT_MATCH = True
+    ROWS_CANNOT_MATCH = False
+
+    def __init__(self, expr, schema, struct, start, end):
+        self.expr = expr
+        self.schema = schema
+        self.struct = struct
+        self.start = start
+        self.end = end
+        self.lower_bounds = None
+        self.upper_bounds = None
+        self.midpoint = None
+
+    def eval(self, row_group):
+        """
+        Returns a boolean that determines if the given row-group may contain rows
+        for the assigned predicate and read-range[start, end]
+
+        Parameters
+        ----------
+        row_group : fastparquet.parquet_thrift.parquet.ttypes.RowGroup
```

Comment:
This information is exposed in `pyarrow` too, see https://github.com/JDASoftwareGroup/kartothek/blob/master/kartothek/serialization/_parquet.py for an implementation with predicate pushdown using only `pyarrow`.

Reply:
Thanks for pointing me to this. Much appreciated.

Reply:
@xhochy I'm seeing some cases where the pyarrow statistics are `None` but the fastparquet stats have them. 

I took a quick look through the arrow JIRAs and did not see anything listed.  Have you ever seen this before?

Reply:
Do you have an example? I know that in some edge cases, e.g. where a datetime column is all null that no statistics are set but otherwise this is normally not the case.

Reply:
This specific example I hit looks as follows:
running `parquet meta` yields the following row for the column in question:
country_iso_code               BINARY    G _ R     190099    0.00 B     0       "AE" / "AR"

fastparquet statistics object looks like this:
```
distinct_count = {NoneType} None
max = {bytes} b'AE'
max_value = {NoneType} None
min = {bytes} b'AR'
min_value = {NoneType} None
null_count = {int} 0
```

pyarrow stats are None, ColumnChunkMD looks like
```
compression = {str} 'GZIP'
data_page_offset = {int} 2129453
dictionary_page_offset = {NoneType} None
encodings = {tuple} <class 'tuple'>: ('RLE', 'PLAIN_DICTIONARY', 'BIT_PACKED')
file_offset = {int} 2129453
file_path = {str} ''
has_dictionary_page = {bool} False
has_index_page = {str} <err_string>
index_page_offset = {str} <err string>
is_stats_set = {bool} False
num_values = {int} 190099
path_in_schema = {str} 'country_iso_code'
physical_type = {str} 'BYTE_ARRAY'
statistics = {NoneType} None
total_compressed_size = {int} 214
total_uncompressed_size = {int} 180
```

I can open a JIRA on arrow if I'm not missing something.

Reply:
Did you create this file with `parquet-mr`? If so, which version? We disable the loading for statistics for BYTE_ARRAY columns that were created with versions before 1.10 as these were wrong. Given that `max_value` is not set in `fastparquet`, I guess you have such a file.

You can see in `cpp/src/parquet/statistics_test.cc` that we explicitly test for that: https://github.com/apache/arrow/blob/558263f51514fb9f87d5b1af262e5a0352f97b64/cpp/src/parquet/statistics_test.cc#L529-L563

cc @rdblue who may be better aware of the sort ordering things in combination with Netflix-generated Parquet files ;)

Reply:
yes, written with Spark, it looks like spark is using an internal fork of parquet-mr 1.9.1, but indeed @rdblue probably knows the specifics of our setup better.

Reply:
This is the problem that @xhochy identified. We use a Parquet version based on 1.9, with a fix and an additional version range where we trust the file stats.

I'm going to be rebasing on 1.10.1 soon, which should fix this problem for tests.

Reply:
@rdblue Thanks for confirming!

@TGooch44 Then in this case, `pyarrow` is correct in not returning any statistics. You shouldn't use the one returned by `fastparquet`.